contracts/non-fungible-tokens/andromeda-crowdfund/src/contract.rs

```diff
@@ -288,6 +300,149 @@ fn execute_start_campaign(
+    let ExecuteContext {
+        ref deps, ref info, ..
+    } = ctx;
+
+    // Ensure campaign accepting coin is received
+    let campaign_config = get_config(deps.storage)?;
+    ensure!(
+        info.funds.len() == 1,
+        ContractError::InvalidFunds {
+            msg: format!(
+                "Only {} is accepted by the campaign.",
+                campaign_config.denom
+            ),
+        }
+    );
+
+    let payment: &Coin = &info.funds[0];
+
+    let sender = info.sender.to_string();
+    let denom = payment.denom.clone();
+    let amount = payment.amount;
+    purchase_tiers(ctx, Asset::NativeToken(denom), amount, sender, orders)
+}
+
+fn handle_receive_cw20(
+    ctx: ExecuteContext,
+    cw20_msg: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let ExecuteContext { ref info, .. } = ctx;
+
+    let amount = cw20_msg.amount;
+    let sender = cw20_msg.sender;
+    let denom = AndrAddr::from_string(info.sender.clone());
+    match from_json(&cw20_msg.msg)? {
+        Cw20HookMsg::PurchaseTiers { orders } => {
+            purchase_tiers(ctx, Asset::Cw20Token(denom), amount, sender, orders)
+        }
+    }
+}
+
+fn purchase_tiers(
+    ctx: ExecuteContext,
+    denom: Asset,
+    amount: Uint128,
+    sender: String,
+    orders: Vec<SimpleTierOrder>,
+) -> Result<Response, ContractError> {
+    let ExecuteContext { deps, env, .. } = ctx;
+
+    let campaign_config = get_config(deps.storage)?;
+    let mut current_cap = get_current_cap(deps.storage);
+
+    let current_stage = get_current_stage(deps.storage);
+
+    // Tier can be purchased on ONGOING stage
+    ensure!(
+        current_stage == CampaignStage::ONGOING,
+        ContractError::InvalidCampaignOperation {
+            operation: "purchase_tiers".to_string(),
+            stage: current_stage.to_string()
+        }
+    );
+
+    // Need to wait until start_time
+    ensure!(
+        campaign_config
+            .start_time
+            .unwrap_or_default()
+            .is_expired(&env.block),
+        ContractError::CampaignNotStarted {}
+    );
+
+    // Campaign is expired or should be ended due to overfunding
+    ensure!(
+        !campaign_config.end_time.is_expired(&env.block)
+            || campaign_config.hard_cap.unwrap_or(current_cap) > current_cap,
+        ContractError::CampaignEnded {}
+    );
+
+    // Ensure campaign accepting coin is received
+    ensure!(
+        denom == campaign_config.denom,
+        ContractError::InvalidFunds {
+            msg: format!(
+                "Only {} is accepted by the campaign.",
+                campaign_config.denom
+            ),
+        }
+    );
+
+    let mut full_orders = Vec::<TierOrder>::new();
+
+    // Measure the total cost for orders
+    let total_cost = orders.iter().try_fold(Uint128::zero(), |sum, order| {
+        let tier = get_tier(deps.storage, u64::from(order.level))?;
+        let uint128: Result<Uint128, ContractError> = Ok(sum + tier.price * order.amount);
```

Comment:
nit: rename to `new_sum` or something similar.
```suggestion
        let new_sum: Result<Uint128, ContractError> = Ok(sum + tier.price * order.amount);
```